src/backends/torch/torchdataset.cc

```diff
@@ -533,52 +546,73 @@ namespace dd
-                    else if (_segmentation)
-                      _img_rand_aug_cv.augment_with_segmap(bgr, bw_target);
-                    else
-                      _img_rand_aug_cv.augment(bgr);
-                  }
-                else
-                  {
-                    // cropping requires test set 'augmentation'
-                    if (_bbox)
+                    cv::Mat bgr_tmp = bgr.clone();
+                    cv::Mat bw_target_tmp;
+                    std::vector<torch::Tensor> t_tmp;
+                    if (_segmentation)
+                      bw_target_tmp = bw_target.clone();
+                    else if (_bbox)
+                      t_tmp = t;
+
+                    // data augmentation can apply here, with OpenCV
+                    if (!_test)
                       {
-                        _img_rand_aug_cv.augment_test_with_bbox(bgr, t);
+                        if (_bbox)
+                          _img_rand_aug_cv.augment_with_bbox(bgr, t);
+                        else if (_segmentation)
+                          _img_rand_aug_cv.augment_with_segmap(bgr, bw_target);
+                        else
+                          _img_rand_aug_cv.augment(bgr);
                       }
-                    else if (_segmentation)
-                      _img_rand_aug_cv.augment_test_with_segmap(bgr,
-                                                                bw_target);
                     else
-                      _img_rand_aug_cv.augment_test(bgr);
-                  }
+                      {
+                        // cropping requires test set 'augmentation'
+                        if (_bbox)
+                          {
+                            _img_rand_aug_cv.augment_test_with_bbox(bgr, t);
+                          }
+                        else if (_segmentation)
+                          _img_rand_aug_cv.augment_test_with_segmap(bgr,
+                                                                    bw_target);
+                        else
+                          _img_rand_aug_cv.augment_test(bgr);
+                      }
 
-                torch::Tensor imgt = image_to_tensor(bgr, bgr.rows, bgr.cols);
-                d.push_back(imgt);
+                    torch::Tensor imgt
+                        = image_to_tensor(bgr, bgr.rows, bgr.cols);
+                    d.push_back(imgt);
 
-                if (_segmentation)
-                  {
-                    at::Tensor targett_seg = image_to_tensor(
-                        bw_target, bw_target.rows, bw_target.cols, true);
-                    t.push_back(targett_seg);
-                  }
-              }
+                    if (_segmentation)
+                      {
+                        at::Tensor targett_seg = image_to_tensor(
+                            bw_target, bw_target.rows, bw_target.cols, true);
+                        t.push_back(targett_seg);
+                        bw_target = bw_target_tmp;
```

Comment:
Better to do the other way around? Work on bw_target_tmp, so you don't have to replace bw_target in the end.
You can also rename `bw_target_tmp` -> `bw_target_sample`

Reply:
I applied the name change only.